dask/dataframe/io/parquet/fastparquet.py

```diff
@@ -759,31 +814,67 @@ def read_partition(cls, fs, piece, columns, index, categories=(), **kwargs):
+                # We have paths to read from
                 assert parquet_file is None
+
+                row_groups = []
+                rg_offset = 0
                 parquet_file = ParquetFile(
-                    piece[0], open_with=fs.open, sep=fs.sep, **kwargs.get("file", {})
+                    [p[0] for p in pieces],
+                    open_with=fs.open,
+                    sep=fs.sep,
+                    **kwargs.get("file", {}),
                 )
-                rg_indices = piece[1] or list(range(len(parquet_file.row_groups)))
+                for piece in pieces:
+                    _pf = (
+                        parquet_file
+                        if len(pieces) == 1
+                        else ParquetFile(
+                            piece[0],
+                            open_with=fs.open,
+                            sep=fs.sep,
+                            **kwargs.get("file", {}),
+                        )
+                    )
+                    n_local_row_groups = len(_pf.row_groups)
+                    local_rg_indices = piece[1] or list(range(n_local_row_groups))
+                    row_groups += [
+                        parquet_file.row_groups[rg + rg_offset]
+                        for rg in local_rg_indices
+                    ]
+                    rg_offset += n_local_row_groups
+                update_parquet_file = len(row_groups) < len(parquet_file.row_groups)
 
-                # `piece[1]` will contain row-group indices
-                row_groups = [parquet_file.row_groups[rg] for rg in rg_indices]
             elif parquet_file:
-                # `piece[1]` will contain actual row-group objects,
-                # but they may be pickled
-                row_groups = piece[0]
-                if isinstance(row_groups, bytes):
-                    row_groups = pickle.loads(row_groups)
-                parquet_file.fmd.row_groups = row_groups
-                # NOTE: May lose cats after `_set_attrs` call
-                save_cats = parquet_file.cats
-                parquet_file._set_attrs()
-                parquet_file.cats = save_cats
+
+                row_groups = []
+                for piece in pieces:
+                    # `piece[1]` will contain actual row-group objects,
+                    # but they may be pickled
+                    rgs = piece[0]
+                    if isinstance(rgs, bytes):
+                        rgs = pickle.loads(rgs)
+                    row_groups += rgs
+                update_parquet_file = True
+
             else:
                 raise ValueError("Neither path nor ParquetFile detected!")
 
+            if update_parquet_file:
+                with _FP_FILE_LOCK:
+                    parquet_file.fmd.row_groups = row_groups
+                    # NOTE: May lose cats after `_set_attrs` call
+                    save_cats = parquet_file.cats
+                    parquet_file._set_attrs()
+                    parquet_file.cats = save_cats
```

Comment:
@martindurant - Can you clarify the `copy` change you are suggesting here?  I am not understandting the correct way to do this.

Reply:
I am suggesting `parquet_file = copy.copy(parquet_file)` above this block, and no lock. That might be enough, or maybe you would also need `parquet_file.fmd = copy.copy(parquet_file.fmd)`. Both of those would be very fast operations.

Reply:
Did you try both together, with copy.copy(parquet_file) first?

Reply:
Right - I also tried this:

```python
                parquet_file = copy.copy(parquet_file)
                parquet_file.fmd = copy.copy(parquet_file.fmd)
                parquet_file.fmd.row_groups = row_groups
                # NOTE: May lose cats after `_set_attrs` call
                save_cats = parquet_file.cats
                parquet_file._set_attrs()
                parquet_file.cats = save_cats
```

Reply:
OK, so if that wasn't good, then we can stick with the lock, but I am a bit puzzled.

Reply:
>I am a bit puzzled

likewise